sycl/doc/extensions/MaxWorkGroupQueries/max_work_group_query.md

```diff
@@ -0,0 +1,81 @@
+
+## Examples
+
+```c++
+sycl::device gpu = sycl::device{sycl::gpu_selector{}};
+std::cout << gpu.get_info<sycl::info::device::name>() << '\n';
+
+#ifdef SYCL_EXT_ONEAPI_MAX_WORK_GROUP_QUERY
+sycl::id<3> groups = gpu.get_info<sycl::info::device::ext_oneapi_max_number_work_groups>();
+size_t global_groups = gpu.get_info<sycl::info::device::ext_oneapi_max_global_number_work_groups>();
+std::cout << "Max number groups: x_max: " << groups[2] << " y_max: " << groups[1] << " z_max: " << groups[0] << '\n';
+std::cout << "Max global number groups: " << global_groups << '\n';
+#endif
+```
+
+Ouputs to the console:
+
+```
+NVIDIA ...
+Max number groups: x_max: 2147483647 y_max: 65535 z_max: 65535
+Max global number groups: 2147483647
+```
+
+See: [CUDA Toolkit Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities)
+
+Then the following `assert` should be met at kernel submission:
+
+```C++
+sycl::nd_range<3> work_range(global_size, local_size);
+
+assert(global_size[2] <= groups[2]
+    && global_size[1] <= groups[1]
+    && global_size[0] <= groups[0]);
+
+assert(global_size[2] * global_size[1] * global_size[0] <= global_groups); //Make sure not to exceed integer representation size in the multiplication.
+
+gpu_queue.submit(work_range, ...);
+```
+
+## Implementation
+
+### Templated queries
+
+Right now, DPC++ does not support templated device descriptors as they are defined in the SYCL specification section 4.6.4.2 "Device information descriptors". When the implementation will support this syntax, `ext_oneapi_max_number_work_groups` should be replaced by the three following descriptors:
+
+| Device descriptors                                 | Return type | Description                                                                                                                                                                                                          |
+| -------------------------------------------------- | ----------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| info::device::ext_oneapi_max_number_work_groups<1> |  id<1>      | Returns the maximum number of work-groups that can be submitted in each dimension of the `globalSize` of a `nd_range`. The minimum value is `(1)` if the device is different than `info::device_type::custom`.       |
+| info::device::ext_oneapi_max_number_work_groups<2> |  id<2>      | Returns the maximum number of work-groups that can be submitted in each dimension of the `globalSize` of a `nd_range`. The minimum value is `(1, 1)` if the device is different than `info::device_type::custom`.    |
```

Comment:
Just a nit on the wording here:

> Returns the maximum number of work-groups that can be submitted in each dimension of the `globalSize` for a 2-dimensional `nd_range` kernel. The minimum value is `(1, 1)` if the device is different than `info::device_type::custom`.

This makes it clear that the query is only valid for a 2-dimensional kernel.

If you decide to add the three queries `ext_oneapi_max_number_work_groups_1d`, etc., then you can use similar wording there.

Reply:
I specified it concerns only a `nd_range<[1,2,3]>`